File: SearchA2DMatrixII/Solution.py
class Solution(object):
    def searchRow(self, row, target):
        l = 0
        r = len(row)
        while l < r:
            m = l + (r - l) // 2
            if row[m] < target:
                l = m + 1
            elif row[m] == target:
                return True
            else:
                r = m
        return False
        
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        l = 0
        r = len(matrix)
        while l < r:
            m = l + (r - l) // 2
            if matrix[m][0] <= target:
                l = m + 1
            else:
                r = m
        i = l - 1
        while i >= 0:
            if self.searchRow(matrix[i], target) == True:
                return True
            i -= 1
        return False

File: SearchA2DMatrixII/Solution.java
public class Solution {
    private boolean searchRow(int[] row, int target) {
        int l = 0;
        int r = row.length;
        int m = 0;
        while (l < r) {
            m = l + (r - l) / 2;
            if (row[m] < target) {
                l = m + 1;
            } else if (row[m] == target) {
                return true;
            } else {
                r = m;
            }
        }
        return false;
    }

    public boolean searchMatrix(int[][] matrix, int target) {
        int l = 0;
        int r = matrix.length;
        int m = 0;
        while (l < r) {
            m = l + (r - l) / 2;
            if (matrix[m][0] <= target) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        for (int i = l - 1; i >= 0; i--) {
            if (searchRow(matrix[i], target)) return true;
        }
        return false;
    }
}
